language/move-lang/src/to_bytecode/translate.rs

```diff
@@ -0,0 +1,1022 @@
+    lvalues_(context, function_frame, code, ls.into_iter())
+}
+
+fn lvalues_(
+    context: &mut Context,
+    function_frame: &mut FunctionFrame,
+    code: &mut Code,
+    ls: impl std::iter::DoubleEndedIterator<Item = G::LValue>,
+) -> Result<()> {
+    for l in ls.rev() {
+        lvalue(context, function_frame, code, l)?
+    }
+    Ok(())
+}
+
+fn lvalue(
+    context: &mut Context,
+    function_frame: &mut FunctionFrame,
+    code: &mut Code,
+    sp!(loc, l_): G::LValue,
+) -> Result<()> {
+    use F::Bytecode as B;
+    use G::LValue_ as L;
+    match l_ {
+        L::Ignore => {
+            code.push(B::Pop);
+            function_frame.pop();
+        }
+        L::Var(v, _) => {
+            let loc_idx = function_frame.local(&v);
+            code.push(B::StLoc(loc_idx));
+            function_frame.pop();
+        }
+        L::Unpack(s, tys, field_ls) => {
+            let def_idx = context.struct_definition_index(&s);
+            let local_sig = F::LocalsSignature(base_types(context, tys)?);
+            let local_idx = context.locals_signature_index(loc, local_sig)?;
+
+            function_frame.pop();
+            code.push(B::Unpack(def_idx, local_idx));
+            function_frame.pushn(loc, field_ls.len())?;
+
+            lvalues_(
+                context,
+                function_frame,
+                code,
+                field_ls.into_iter().map(|(_, l)| l),
+            )?;
+        }
+    }
+    Ok(())
+}
+
+//**************************************************************************************************
+// Expressions
+//**************************************************************************************************
+
+fn exp(
+    context: &mut Context,
+    function_frame: &mut FunctionFrame,
+    code: &mut Code,
+    e: Box<G::Exp>,
+) -> Result<()> {
+    exp_(context, function_frame, code, *e)
+}
+
+fn exp_(
+    context: &mut Context,
+    function_frame: &mut FunctionFrame,
+    code: &mut Code,
+    e: G::Exp,
+) -> Result<()> {
+    use Value_ as V;
+    use F::Bytecode as B;
+    use G::UnannotatedExp_ as E;
+    let ty = e.ty;
+    let sp!(eloc, e_) = e.exp;
+    match e_ {
+        E::UnresolvedError => panic!("ICE should not have reached compilation if there are errors"),
+        E::Unit => (),
+        E::Value(sp!(_, V::Address(a))) => {
+            let idx = context.address_index(eloc, a)?;
+            code.push(B::LdAddr(idx));
+            function_frame.push(eloc)?;
+        }
+        E::Value(sp!(_, V::U64(u))) => {
+            code.push(B::LdConst(u));
+            function_frame.push(eloc)?;
+        }
+        E::Value(sp!(_, V::Bool(b))) => {
+            code.push(if b { B::LdTrue } else { B::LdFalse });
+            function_frame.push(eloc)?;
+        }
+        E::Value(sp!(_, V::Bytearray(bytes))) => {
+            let idx = context.byte_array_index(eloc, bytes)?;
+            code.push(B::LdByteArray(idx));
+            function_frame.push(eloc)?;
+        }
```

Comment:
What are your thoughts on changing this to a top-level match along with an inner match to case on the value type? I'm just thinking that it might make it a bit more readable.